test/integration/provider/admin/account/logos_controller_test.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class Provider::Admin::Account::LogosControllerTest < ActionDispatch::IntegrationTest
+  include ActiveSupport::Testing::FileFixtures
+
+  self.file_fixture_path = Rails.root.join('test', 'fixtures')
+
+  def setup
+    @provider = FactoryBot.create(:provider_account)
+    provider.settings.allow_branding!
+    login! provider
+  end
+
+  attr_reader :provider
+
+  test 'update when it should work' do
+    assert_change(of: -> { provider.profile.reload.logo_file_name }, from: nil, to: 'hypnotoad.jpg') do
+      put provider_admin_account_logo_path, {profile: {logo: logo_file}}
+    end
+
+    assert_redirected_to edit_provider_admin_account_logo_path
+    assert_equal 'Logo was successfully created.', flash[:notice]
+    assert_nil flash[:error]
+  end
+
+  test 'update when it should fail' do
+    assert_no_change(of: -> { provider.profile.reload.logo_file_name }) do
+      put provider_admin_account_logo_path, {profile: {logo: countries_yaml_file}}
+    end
+
+    assert_redirected_to edit_provider_admin_account_logo_path
+    assert_match /Logo content type invalid/, flash[:error]
+    assert_nil flash[:notice]
+  end
+
+  test 'update unauthorized without branding switch' do
+    provider.settings.deny_branding!
+
+    assert_no_change(of: -> { provider.profile.reload.logo_file_name }) do
+      put provider_admin_account_logo_path, {profile: {logo: logo_file}}
+    end
+
+    assert_response :forbidden
+  end
+
+  test 'destroy when it should work' do
+    provider.profile.update_attribute(:logo, Rack::Test::UploadedFile.new(logo_file, 'image/jpeg', true))
+
+    assert_change(of: -> { provider.profile.reload.logo_file_name }, from: provider.profile.logo_file_name, to: nil) do
+      delete provider_admin_account_logo_path, {profile: {logo: logo_file}}
+    end
+
+    assert_redirected_to edit_provider_admin_account_logo_path
+    assert_equal 'Your logo was successfully deleted.', flash[:notice]
+    assert_nil flash[:error]
+  end
+
+  test 'destroy when it should fail' do
```

Comment:
I do not get this test.
Can't you provide pre-condition on when it will fail? The stubs are strange to me

Reply:
@hallelujah The controller before this test was skipping validations and this tests means to ensure that this doesn't happen and if any validation makes it fail, then it should fail.
Right now there is no validation that would make it fail and that's why I've done it like this.

Reply:
So what should I do about this? @hallelujah

Reply:
can you try to upload not valid file? should be the same as stubbing `valid?: false`.. right?

Reply:
@macejmic That's what I did for the `update` (see https://github.com/3scale/porta/pull/975/files#diff-cdaeffaabb0811cb7c29c744223afd65R28) but I can't do that for the `destroy` because it only takes `nil` and not the params (see https://github.com/3scale/porta/pull/975/files#diff-5512c93ae956a903f6808306318f5dc3R14)

Reply:
i got it now.. you're testing everything! 😺 i'm ok with that..